.zuul.playbooks/playbooks/tox/main.yaml

```diff
@@ -0,0 +1,26 @@
+- hosts: all
+  tasks:
+
+    - name: Install tox
+      include_role:
+        name: ensure-tox
+
+    # TODO(ianw) : let's see if this is interferring
+    # with some of the pip tox runs
+    - name: Remove OpenDev mirror config
+      file:
+        path: /etc/pip.conf
+        state: absent
+      become: yes
+
+    - name: Run tox unit tests
+      include_role:
+        name: tox
+      vars:
+        tox_extra_args: "-- -m unit -n auto --use-venv"
+
+    - name: Run tox integration tests with no network
+      include_role:
+        name: tox
+      vars:
+        tox_extra_args: '-- -m integration -n auto --use-venv'
```

Comment:
Can units and integrations run in parallel jobs? Would it make sense?

Reply:
They can.  The other existing jobs seem to run separately so it fails earlier.  Personally I'd just run it all as you suggest because that way you see all failures your change introduces, without having to run it through multiple times

Reply:
Yep. The current CI setup resorts to using hacks with early cancellations because of the limited resources. So if Zuul can handle the load, it's better to build a more responsive setup.

Reply:
removed the unit/integration split

Reply:
Weren't we going to do the opposite?

Reply:
> Can units and integrations run in parallel jobs? Would it make sense?

> removed the unit/integration split

> Weren't we going to do the opposite?

This just has one run of tox that does not specify any limiters hence does unit and integration tests all together and in parallel.  Were you thinking something else?

Reply:
I was thinking that if resources allow, we could have separate jobs (VMs?) spawned in parallel, one runs unit tests and the other runs integration tests. I imagine, this could be provide a responsive experience... But that's probably up to the maintainers to decide if they want this.

cc @pradyunsg

Reply:
I mean, anything is possible, but I don't think it's worth spinning up two instances for this.  It's going to vary slightly per-provider, but the job is running between 4 - 6 minutes overall and the tox run is ~3 minutes.  So there's approximately as much overhead starting the testing environment as running the tests.

Reply:
Fair enough. Is it possible to use pre-baked images to improve the cold start? Like other CIs do.

Reply:
There's always room for improvement, but there's probably two parts to that.  Part of the setup is cloud and host specific things; stuff that we don't actually know until the node starts like what mirror to point it at and ssh keys.

The other bits are setting up tox and pip.  We have very generic jobs because these are widely portable -- people use these building blocks not just on the VM images Opendev builds, but their own images in openshift, k8s, google compute, aws etc etc.  For that reason, a lot of the framework in https://zuul-ci.org/docs/zuul-jobs/ is very intentionally lowest-common-denominator.  

It's a trade off -- other CI systems don't have goals of making a completely re-usable environment you can roll out to a wide range of very minimal testing environments.  So we're always considering that when we build bits of the jobs.

Reply:
I understand that. But let me rephrase: how hard would it be to make some image from say quay.io run this build?

Reply:
In the way that I think you mean, this isn't possible in Opendev as the Opendev node environment is not container-based.  Note that Zuul can certainly be setup to provide resources in many different ways, but in Opendev's Zuul instance you are getting a virtual-machine from the images we pre-build daily (qcow images are available @ https://nb01.opendev.org/images/ and I can go into how and why they are built, but I don't think that's the point here).  So you can't tell this job to run on an arbitrary container.

You *can* install docker/podman on the host, and do things inside that like run jobs in whatever container you want.  For example, the pypa manylinux builds use this to build wheels in a container; e.g. https://github.com/pyca/cryptography/blob/master/.zuul.playbooks/playbooks/wheel/roles/build-wheel-manylinux/tasks/main.yaml

This is a flexibility win, but because you have to install docker and pull the image, not a speed win for running something like tox.

Reply:
Got it, thanks! I was just thinking about what we could do in the context of what @pradyunsg wished for when saying that it better be fast.